src/lib/tftp/tftp.c

```diff
@@ -0,0 +1,403 @@
+	return 0;
+}
+
+int tftp_resend_cmd(struct tftp_stream *s) {
+	if (0 != tftp_msg_send(&s->snd, s->snd_len, s)) {
+		return -1;
+	}
+
+	if (0 != tftp_msg_recv(&s->rcv, &s->rcv_len, s)) {
+		return -2;
+	}
+
+	return 0;
+}
+
+struct tftp_stream *tftp_new_stream(const char *host, const char *file, int dir, bool binary_mode) {
+	struct tftp_stream *s = &stream;
+
+	memset(s, 0, sizeof(*s));
+
+	if (0 != make_remote_addr(host,
+				&s->rem_addr,
+				&s->rem_addrlen)) {
+		goto fail;
+	}
+	if (-1 == (s->sock = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP))) {
+		goto fail;
+	}
+
+	if (dir == TFTP_DIR_PUT) {
+		if (-1 == connect(s->sock,
+				(struct sockaddr *) &s->rem_addr,
+				s->rem_addrlen)) {
+			goto fail;
+		}
+
+
+		if (0 != tftp_build_msg_cmd(&s->snd,
+				&s->snd_len,
+				TFTP_WRQ,
+				file,
+				get_transfer_mode(binary_mode))) {
+			goto fail;
+		}
+
+	} else {
+		if (0 != tftp_build_msg_cmd(&s->snd,
+					&s->snd_len,
+					TFTP_RRQ,
+					file,
+					get_transfer_mode(binary_mode))) {
+			goto fail;
+		}
+	}
+
+	do {
+		if (tftp_resend_cmd(s)) {
+			printk("ERROR\n");
+			break;
+		}
+	} while (!msg_with_correct_len(&s->rcv, s->rcv_len));
+
+	strncpy(s->filename, file, sizeof(s->filename));
+
+	return s;
+
+fail:
+	if (s->sock >= 0) {
+		close(s->sock);
+	}
+
+	memset(s, 0, sizeof(*s));
+
+	return NULL;
+}
+
+int tftp_delete_stream(struct tftp_stream *s) {
+	if (s == NULL) {
+		return -EINVAL;
+	}
+
+	if (s->sock >= 0) {
+		close(s->sock);
+	}
+
+	memset(s, 0, sizeof(*s));
+
+	return 0;
+}
+
+int tftp_stream_write(struct tftp_stream *s, uint8_t *buf, size_t len) {
+	if (s == NULL) {
+		log_warning("stream is NULL, do nothing");
+		return 0;
+	}
+
+	switch (ntohs(s->rcv.opcode)) {
+		case TFTP_ACK:
+			while (ntohs(s->rcv.op.ack.block_num) != s->pkg_number) {
+				tftp_resend_cmd(s); /* invalid acknowledgement, send again */
+			}
+			break;
+		case TFTP_ERROR:
+			return -s->rcv.op.err.error_code;
+		default:
+			/* error */
+			return 0;
+	}
+
+	/* whether we have more data to transfer? */
+	if ((s->pkg_number != 0) && (s->snd_len != sizeof s->snd)) {
+		/* end of transmission */
+	}
+
+	/* get next stuff of data */
+	if (0 != tftp_build_msg_data(&s->snd, &s->snd_len, buf, ++s->pkg_number)) {
+		goto fail;
+	}
+
+	/* send request / data */
+	if (0 != tftp_msg_send(&s->snd, s->snd_len, s)) {
+		goto fail;
+	}
+
+	return s->snd_len;
+fail:
+	return 0;
+}
+
+int tftp_stream_read(struct tftp_stream *s, uint8_t *buf) {
+	int data_len;
+
+	if (s->transmission_end) {
+		return 0;
+	}
+
+	switch (ntohs(s->rcv.opcode)) {
+		case TFTP_DATA:
+			while (ntohs(s->rcv.op.ack.block_num) != s->pkg_number + 1) {
+				tftp_resend_cmd(s);
+			}
+
+			data_len = s->rcv_len - (sizeof s->rcv - sizeof s->rcv.op.data.stuff);
+			memcpy(buf, s->rcv.op.data.stuff, data_len);
+
+			if (s->rcv_len != sizeof(s->rcv)) {
+				s->transmission_end = true;
+			}
+			break;
+		case TFTP_ERROR:
+			return -s->rcv.op.err.error_code;
+			goto fail;
+		default:
+			while(1);
```

Comment:
while(1) looks very unsafe. Is there another way to handle error? Probably `goto fail` or at least print something.

Reply:
Fixed